src/test/scala/org/apache/spark/sql/delta/DeltaSuite.scala

```diff
@@ -198,40 +198,81 @@ class DeltaSuite extends QueryTest
         .mode("overwrite")
-        .option(DeltaOptions.REPLACE_WHERE_OPTION, "not_a_column = true")
+        .option(DeltaOptions.REPLACE_WHERE_OPTION, "")
         .save(tempDir.toString)
     }.getMessage
-    assert(e3 == "Predicate references non-partition column 'not_a_column'. Only the " +
-      "partition columns may be referenced: [is_odd];")
+    assert(e5.contains("Cannot recognize the predicate ''"))
+  }
 
-    var e4 = intercept[AnalysisException] {
-      Seq(6).toDF()
-        .withColumn("is_odd", $"value" % 2 =!= 0)
+  test("replaceWhere with multiple non-partition predicates") {
+    withTempDir { tempDir =>
+      def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
+
+      Seq((1, 10), (2, 20), (3, 10), (4, 20), (4, 10)).toDF("key", "value")
+        .write
+        .format("delta")
+        .partitionBy("value")
+        .save(tempDir.getCanonicalPath)
+
+      Seq((4, 30), (5, 20), (6, 70)).toDF("key", "value")
         .write
         .format("delta")
         .mode("overwrite")
-        .option(DeltaOptions.REPLACE_WHERE_OPTION, "value = 1")
-        .save(tempDir.toString)
-    }.getMessage
-    assert(e4 == "Predicate references non-partition column 'value'. Only the " +
-      "partition columns may be referenced: [is_odd];")
+        .option(DeltaOptions.REPLACE_WHERE_OPTION, "key = 4 OR key = 5")
+        .save(tempDir.getCanonicalPath)
 
-    val e5 = intercept[AnalysisException] {
-      Seq(6).toDF()
-        .withColumn("is_odd", $"value" % 2 =!= 0)
+      checkDatasetUnorderly(data.toDF.as[(Int, Int)],
+        1 -> 10, 3 -> 10, 2 -> 20, 5 -> 20, 4 -> 30, 6 -> 70)
+    }
+  }
+
+  test("replaceWhere with nested predicates") {
+    withTempDir { tempDir =>
+      def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
+
+      Seq((1, 10), (2, 20), (3, 10), (4, 20), (4, 10)).toDF("key", "value")
+        .write
+        .format("delta")
+        .partitionBy("value")
+        .save(tempDir.getCanonicalPath)
+
+      Seq((4, 30), (5, 20)).toDF("key", "value")
         .write
         .format("delta")
         .mode("overwrite")
-        .option(DeltaOptions.REPLACE_WHERE_OPTION, "")
-        .save(tempDir.toString)
-    }.getMessage
-    assert(e5.contains("Cannot recognize the predicate ''"))
+        .option(DeltaOptions.REPLACE_WHERE_OPTION, "(key = 4 AND value = 20) OR key = 5")
+        .save(tempDir.getCanonicalPath)
+
+      checkDatasetUnorderly(data.toDF.as[(Int, Int)],
+        1 -> 10, 3 -> 10, 4 -> 10, 2 -> 20, 5 -> 20, 4 -> 30)
+    }
+  }
+
+  test("replaceWhere with no matching predicates") {
+    withTempDir { tempDir =>
+      def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
+
+      Seq((1, 10), (2, 20), (3, 10), (4, 20), (4, 10)).toDF("key", "value")
+        .write
+        .format("delta")
+        .partitionBy("value")
+        .save(tempDir.getCanonicalPath)
+
+      Seq((4, 30), (5, 20)).toDF("key", "value")
```

Comment:
Yeah, in order to follow the semantics of replaceWhere, this should throw an exception saying that the new data did not contain the replaceWhere condition.